.doom.d/config.org
#+TITLE: Doom Emacs Config
#+AUTHOR: realsontaran
#+STARTUP: content
#+PROPERTY: header-args :tangle config.el

* My Doom Emacs Config
** User
#+begin_src emacs-lisp
(setq user-full-name "realsontaran"
      user-mail-address "therealsontaran@protonmailcom")
#+end_src

** Font, Theme
Setting up nice user interface.
#+begin_src emacs-lisp
(setq doom-font (font-spec :family "JetBrains Mono" :size 14)
      doom-big-font (font-spec :family "JetBrains Mono" :size 20)
      doom-variable-pitch-font (font-spec :family "Overpass" :size 20)
      doom-serif-font (font-spec :family "IBM Plex Mono" :weight 'light))

(setq display-line-numbers-type t)
(setq doom-theme 'zaiste)
(setq fancy-splash-image "~/.doom.d/orangutan.png")
(custom-set-faces!
  '(doom-modeline-buffer-modified :foreground "orange"))

;; Add some transparency
;; (set-frame-parameter (selected-frame) 'alpha '(90 90))
;; (add-to-list 'default-frame-alist '(alpha 90 90))
(windmove-default-keybindings)
#+end_src
** Change Theme
#+begin_src emacs-lisp
;; (defun synchronize-theme ()
;;   (let* ((light-theme 'modus-operandi)
;;          (dark-theme 'modus-vivendi)
;;          (start-time-light-theme 6)
;;          (end-time-light-theme 17)
;;          (hour (string-to-number (substring (current-time-string) 11 13)))
;;          (next-theme (if (member hour (number-sequence start-time-light-theme end-time-light-theme))
;;                          light-theme dark-theme)))
;;     (when (not (equal doom-theme next-theme))
;;       (setq doom-theme next-theme)
;;       (load-theme next-theme))))

;; (run-with-timer 0 900 'synchronize-theme)
#+end_src
** Variable Setting
Setting some values.
#+begin_src emacs-lisp
(setq shell-file-name "/bin/zsh")
(setq-default
 delete-by-moving-to-trash t
 window-combination-resize t)

;; (setq doom-fallback-buffer-name "► Doom"
;;       +doom-dashboard-name "► Doom")

;; Mode-line
(display-time-mode t)
(unless (equal "Battery status not available"
               (battery))
  (display-battery-mode 1))
(setq auth-sources '("~/.authinfo"))
(setq save-interprogram-paste-before-kill nil)
#+end_src
** Some Functions
#+begin_src emacs-lisp
(map! :leader
      :desc "Org-Pomodoro"
      "v" #'org-pomodoro)
(map! :ne "M-/" #'comment-line)
(map! :ne "M-n" #'man)
(map! :leader
      :desc "Dired Jump"
      "[" #'dired-jump)
(map! :leader
      :desc "Org-Roam Insert"
      "r" #'org-roam-capture)

#+end_src
** Treemacs
#+begin_src emacs-lisp
(remove-hook 'doom-load-theme-hook #'doom-themes-treemacs-config)
#+end_src
** Which-Key Pop-Up
#+begin_src emacs-lisp
(setq which-key-idle-delay 0.5)
(setq which-key-allow-multiple-replacements t)
(after! which-key
  (pushnew!
   which-key-replacement-alist
   '(("" . "\\`+?evil[-:]?\\(?:a-\\)?\\(.*\\)") . (nil . "◂\\1"))
   '(("\\`g s" . "\\`evilem--?motion-\\(.*\\)") . (nil . "◃\\1"))
   ))
#+end_src
** ORG-MODE
Setting todo keywords, org-agenda directories and some keymapping for funcional workspace.
#+begin_src emacs-lisp
(require 'org-bullets)
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
(after! org
  (setq org-bullets-bullet-list '("●" "⎉" "⁘" "⁙" "⁜"))
  (setq org-priority-faces '((65 :foreground "#e45649")
                        (66 :foreground "#da8548")
                        (67 :foreground "#0098dd")))

  (setq org-directory "~/Dropbox/org/")
  (setq org-log-done 'time)
  (setq org-agenda-files (directory-files-recursively "~/Dropbox/org/" "\\.org$"))
  (setq org-todo-keywords
        '((sequence "TODO" "|" "DONE" "CNCL")
          (sequence "[ ]" "[-]" "|" "[X]")
          (sequence "LECTURE(l)" "|")
          (sequence "HOMEWORK" "PROJ" "TASK" "|" "DONE")
          (sequence "PLAY" "READ" "WATCH" "ONIT" "|" "DONE")
          ))
  (setq org-todo-keyword-faces
        '(("TODO" . "orange")
          ("HOMWRK" . "orange")
          ("DONE" . "green")
          ("CNCL" . (:foreground "red" :weight bold))
          ("LECTURE" . "purple")

st/config.h
/* See LICENSE file for copyright and license details. */

/*
 * appearance
 *
 * font: see http://freedesktop.org/software/fontconfig/fontconfig-user.html
 */
static char *font = "Mononoki Nerd Font:size=10:antialias=true:autohint=true";
static char *font2[] = {"JoyPixels:pixelsize=10:antialias=true:autohint=true"};
static int borderpx = 2;

/*
 * What program is execed by st depends of these precedence rules:
 * 1: program passed with -e
 * 2: utmp option
 * 3: SHELL environment variable
 * 4: value of shell in /etc/passwd
 * 5: value of shell in config.h
 */
static char *shell = "/bin/sh";
char *utmp = NULL;
char *stty_args = "stty raw pass8 nl -echo -iexten -cstopb 38400";

/* identification sequence returned in DA and DECID */
char *vtiden = "\033[?6c";

/* Kerning / character bounding-box multipliers */
static float cwscale = 1.0;
static float chscale = 1.0;

/*
 * word delimiter string
 *
 * More advanced example: L" `'\"()[]{}"
 */
wchar_t *worddelimiters = L" ";

/* selection timeouts (in milliseconds) */
static unsigned int doubleclicktimeout = 300;
static unsigned int tripleclicktimeout = 600;

/* alt screens */
int allowaltscreen = 1;

/*
 * draw latency range in ms - from new content/keypress/etc until drawing.
 * within this range, st draws when content stops arriving (idle). mostly it's
 * near minlatency, but it waits longer for slow updates to avoid partial draw.
 * low minlatency will tear/flicker more, as it can "detect" idle too early.
 */
static double minlatency = 8;
static double maxlatency = 33;

/*
 * blinking timeout (set to 0 to disable blinking) for the terminal blinking
 * attribute.
 */
static unsigned int blinktimeout = 800;

/*
 * interval (in milliseconds) between each successive call to ximspot. This
 * improves terminal performance while not reducing functionality to those
 * whom need XIM support.
 */
int ximspot_update_interval = 1000;

/*
 * thickness of underline and bar cursors
 */
static unsigned int cursorthickness = 2;

/*
 * 1: render most of the lines/blocks characters without using the font for
 *    perfect alignment between cells (U2500 - U259F except dashes/diagonals).
 *    Bold affects lines thickness if boxdraw_bold is not 0. Italic is ignored.
 * 0: disable (render all U25XX glyphs normally from the font).
 */
const int boxdraw = 1;
const int boxdraw_bold = 1;

/* braille (U28XX):  1: render as adjacent "pixels",  0: use font */
const int boxdraw_braille = 1;

/*
 * bell volume. It must be a value between -100 and 100. Use 0 for disabling
 * it
 */
static int bellvolume = 0;

/* default TERM value */
char *termname = "st-256color";

/*
 * spaces per tab
 *
 * When you are changing this value, don't forget to adapt the »it« value in
 * the st.info and appropriately install the st.info in the environment where
 * you use this st version.
 *
 *	it#$tabspaces,
 *
 * Secondly make sure your kernel is not expanding tabs. When running `stty
 * -a` »tab0« should appear. You can tell the terminal to not expand tabs by
 *  running following command:
 *
 *	stty tabs
 */
unsigned int tabspaces = 8;

/* bg opacity */
float alpha = 0.9;

/* Terminal colors (16 first used in escape sequence) */
const char *colorname[] = {

    /* 8 normal colors */
    [0] = "#282828", /* black  / hard contrast: #1d2021 / soft contrast: #32302f
                      */
    [1] = "#cc241d", /* red     */
    [2] = "#98971a", /* green   */
    [3] = "#d79921", /* yellow  / orange = "#d65d0e" */
    [4] = "#458588", /* blue    */
    [5] = "#b16286", /* magenta */
    [6] = "#689d6a", /* cyan    */
    [7] = "#a89984", /* white   */

    /* 8 bright colors */
    [8] = "#928374",  /* black   */
    [9] = "#fb4934",  /* red     */
    [10] = "#b8bb26", /* green   */
    [11] = "#fabd2f", /* yellow  / orange = "#fe8019" */
    [12] = "#83a598", /* blue    */
    [13] = "#d3869b", /* magenta */
    [14] = "#8ec07c", /* cyan    */
    [15] = "#ebdbb2", /* white   */

    /* special colors */
    [256] = "#282828", /* background */
    [257] = "#a89984", /* foreground */
    [258] = "#ebdbb2", /* cursor */
};

/* Default colors (colorname index)
 * foreground, background, cursor */
unsigned int defaultbg = 0;
unsigned int defaultfg = 257;
unsigned int defaultcs = 258;
unsigned int defaultrcs = 258;

/*
 * https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h4-Functions-using-CSI-_-ordered-by-the-final-character-lparen-s-rparen:CSI-Ps-SP-q.1D81
 * Default style of cursor
 * 0: Blinking block
 * 1: Blinking block (default)
 * 2: Steady block ("█")
 * 3: Blinking underline
 * 4: Steady underline ("_")
 * 5: Blinking bar
 * 6: Steady bar ("|")
 * 7: Blinking st cursor
 * 8: Steady st cursor
 */
static unsigned int cursorstyle = 5;
static Rune stcursor = 0x2603; /* snowman (U+2603) */

/*
 * Default columns and rows numbers
 */

static unsigned int cols = 80;
static unsigned int rows = 24;

/*
 * Default colour and shape of the mouse cursor
 */
static unsigned int mouseshape = XC_xterm;
static unsigned int mousefg = 7;
static unsigned int mousebg = 0;

/*
 * Color used to display font attributes when fontconfig selected a font which
 * doesn't match the ones requested.
 */
static unsigned int defaultattr = 11;

/*
 * Xresources preferences to load at startup
 */
ResourcePref resources[] = {
    {"font", STRING, &font},
    {"fontalt0", STRING, &font2[0]},
    {"color0", STRING, &colorname[0]},
    {"color1", STRING, &colorname[1]},
    {"color2", STRING, &colorname[2]},
    {"color3", STRING, &colorname[3]},
    {"color4", STRING, &colorname[4]},
    {"color5", STRING, &colorname[5]},
    {"color6", STRING, &colorname[6]},
    {"color7", STRING, &colorname[7]},
    {"color8", STRING, &colorname[8]},
    {"color9", STRING, &colorname[9]},
    {"color10", STRING, &colorname[10]},
    {"color11", STRING, &colorname[11]},
    {"color12", STRING, &colorname[12]},
    {"color13", STRING, &colorname[13]},
    {"color14", STRING, &colorname[14]},
    {"color15", STRING, &colorname[15]},
    {"background", STRING, &colorname[258]},
    {"foreground", STRING, &colorname[259]},
    {"cursorColor", STRING, &colorname[256]},
    {"termname", STRING, &termname},
    {"shell", STRING, &shell},
    {"blinktimeout", INTEGER, &blinktimeout},
    {"bellvolume", INTEGER, &bellvolume},
    {"tabspaces", INTEGER, &tabspaces},
    {"borderpx", INTEGER, &borderpx},
    {"cwscale", FLOAT, &cwscale},
    {"chscale", FLOAT, &chscale},
    {"alpha", FLOAT, &alpha},
    {"ximspot_update_interval", INTEGER, &ximspot_update_interval},
};

/*
 * Internal mouse shortcuts.
 * Beware that overloading Button1 will disable the selection.
 */
static MouseShortcut mshortcuts[] = {
    /* button               mask            string */
    {Button4, XK_NO_MOD, "\031"},
    {Button5, XK_NO_MOD, "\005"},
};

/* Internal keyboard shortcuts. */
#define MODKEY Mod1Mask
#define TERMMOD (Mod1Mask | ShiftMask)

MouseKey mkeys[] = {
    /* button               mask            function        argument */
    {Button4, XK_NO_MOD, kscrollup, {.i = 1}},
    {Button5, XK_NO_MOD, kscrolldown, {.i = 1}},
    {Button4, TERMMOD, zoom, {.f = +1}},
    {Button5, TERMMOD, zoom, {.f = -1}},
};

static char *openurlcmd[] = {"/bin/sh", "-c", "st-urlhandler", "externalpipe",
                             NULL};

static char *copyurlcmd[] = {
    "/bin/sh", "-c",
    "tmp=$(sed 's/.*│//g' | tr -d '\n' | grep -aEo "
    "'(((http|https|gopher|gemini|ftp|ftps|git)://"
    "|www\\.)[a-zA-Z0-9.]*[:]?[a-zA-Z0-9./"
    "@$&%?$#=_-~]*)|((magnet:\\?xt=urn:btih:)[a-zA-Z0-9]*)' | uniq | sed "
    "'s/^www./http:\\/\\/www\\./g' ); IFS=; [ ! -z $tmp ] && echo $tmp | dmenu "
    "-i -p 'Copy which url?' -l 10 | tr -d '\n' | xclip -selection clipboard",
    "externalpipe", NULL};

static char *copyoutput[] = {"/bin/sh", "-c", "st-copyout", "externalpipe",
                             NULL};

static Shortcut shortcuts[] = {
    /* mask                 keysym          function        argument */
    {XK_ANY_MOD, XK_Break, sendbreak, {.i = 0}},
    {ControlMask, XK_Print, toggleprinter, {.i = 0}},
    {ShiftMask, XK_Print, printscreen, {.i = 0}},
    {XK_ANY_MOD, XK_Print, printsel, {.i = 0}},
    {TERMMOD, XK_Prior, zoom, {.f = +1}},
    {TERMMOD, XK_Next, zoom, {.f = -1}},
    {MODKEY, XK_Home, zoomreset, {.f = 0}},
    {MODKEY, XK_c, clipcopy, {.i = 0}},
    {ShiftMask, XK_Insert, clippaste, {.i = 0}},
    {MODKEY, XK_v, clippaste, {.i = 0}},
    {XK_ANY_MOD, Button2, selpaste, {.i = 0}},
    {MODKEY, XK_Num_Lock, numlock, {.i = 0}},
    {MODKEY, XK_Control_L, iso14755, {.i = 0}},
    {ShiftMask, XK_Page_Up, kscrollup, {.i = -1}},
    {ShiftMask, XK_Page_Down, kscrolldown, {.i = -1}},
    {MODKEY, XK_Page_Up, kscrollup, {.i = -1}},
    {MODKEY, XK_Page_Down, kscrolldown, {.i = -1}},
    {MODKEY, XK_k, kscrollup, {.i = 1}},
    {MODKEY, XK_j, kscrolldown, {.i = 1}},
    {MODKEY, XK_Up, kscrollup, {.i = 1}},
    {MODKEY, XK_Down, kscrolldown, {.i = 1}},
    {MODKEY, XK_u, kscrollup, {.i = -1}},
    {MODKEY, XK_d, kscrolldown, {.i = -1}},
    {TERMMOD, XK_Up, zoom, {.f = +1}},
    {TERMMOD, XK_Down, zoom, {.f = -1}},
    {TERMMOD, XK_K, zoom, {.f = +1}},
    {TERMMOD, XK_J, zoom, {.f = -1}},
    {TERMMOD, XK_U, zoom, {.f = +2}},
    {TERMMOD, XK_D, zoom, {.f = -2}},
    {MODKEY, XK_l, externalpipe, {.v = openurlcmd}},
    {MODKEY, XK_y, externalpipe, {.v = copyurlcmd}},
    {MODKEY, XK_o, externalpipe, {.v = copyoutput}},
};

/*
 * Special keys (change & recompile st.info accordingly)
 *
 * Mask value:
 * * Use XK_ANY_MOD to match the key no matter modifiers state
 * * Use XK_NO_MOD to match the key alone (no modifiers)
 * appkey value:
 * * 0: no value
 * * > 0: keypad application mode enabled
 * *   = 2: term.numlock = 1
 * * < 0: keypad application mode disabled
 * appcursor value:
 * * 0: no value
 * * > 0: cursor application mode enabled
 * * < 0: cursor application mode disabled
 * crlf value
 * * 0: no value
 * * > 0: crlf mode is enabled
 * * < 0: crlf mode is disabled
 *
 * Be careful with the order of the definitions because st searches in
 * this table sequentially, so any XK_ANY_MOD must be in the last
 * position for a key.
 */

/*
 * If you want keys other than the X11 function keys (0xFD00 - 0xFFFF)
 * to be mapped below, add them to this array.
 */
static KeySym mappedkeys[] = {-1};

/*
 * State bits to ignore when matching key or button events.  By default,
 * numlock (Mod2Mask) and keyboard layout (XK_SWITCH_MOD) are ignored.
 */
static uint ignoremod = Mod2Mask | XK_SWITCH_MOD;

/*
 * Override mouse-select while mask is active (when MODE_MOUSE is set).
 * Note that if you want to use ShiftMask with selmasks, set this to an other
 * modifier, set to 0 to not use it.
 */
static uint forceselmod = ShiftMask;

/*
 * This is the huge key array which defines all compatibility to the Linux
 * world. Please decide about changes wisely.
 */
static Key key[] = {
    /* keysym           mask            string      appkey appcursor */
    {XK_KP_Home, ShiftMask, "\033[2J", 0, -1},
    {XK_KP_Home, ShiftMask, "\033[1;2H", 0, +1},
    {XK_KP_Home, XK_ANY_MOD, "\033[H", 0, -1},
    {XK_KP_Home, XK_ANY_MOD, "\033[1~", 0, +1},
    {XK_KP_Up, XK_ANY_MOD, "\033Ox", +1, 0},
    {XK_KP_Up, XK_ANY_MOD, "\033[A", 0, -1},
    {XK_KP_Up, XK_ANY_MOD, "\033OA", 0, +1},
    {XK_KP_Down, XK_ANY_MOD, "\033Or", +1, 0},
    {XK_KP_Down, XK_ANY_MOD, "\033[B", 0, -1},
    {XK_KP_Down, XK_ANY_MOD, "\033OB", 0, +1},
    {XK_KP_Left, XK_ANY_MOD, "\033Ot", +1, 0},
    {XK_KP_Left, XK_ANY_MOD, "\033[D", 0, -1},
    {XK_KP_Left, XK_ANY_MOD, "\033OD", 0, +1},
    {XK_KP_Right, XK_ANY_MOD, "\033Ov", +1, 0},
    {XK_KP_Right, XK_ANY_MOD, "\033[C", 0, -1},
    {XK_KP_Right, XK_ANY_MOD, "\033OC", 0, +1},
    {XK_KP_Prior, ShiftMask, "\033[5;2~", 0, 0},
    {XK_KP_Prior, XK_ANY_MOD, "\033[5~", 0, 0},
    {XK_KP_Begin, XK_ANY_MOD, "\033[E", 0, 0},
    {XK_KP_End, ControlMask, "\033[J", -1, 0},
    {XK_KP_End, ControlMask, "\033[1;5F", +1, 0},
    {XK_KP_End, ShiftMask, "\033[K", -1, 0},
    {XK_KP_End, ShiftMask, "\033[1;2F", +1, 0},
    {XK_KP_End, XK_ANY_MOD, "\033[4~", 0, 0},
    {XK_KP_Next, ShiftMask, "\033[6;2~", 0, 0},
    {XK_KP_Next, XK_ANY_MOD, "\033[6~", 0, 0},
    {XK_KP_Insert, ShiftMask, "\033[2;2~", +1, 0},
    {XK_KP_Insert, ShiftMask, "\033[4l", -1, 0},
    {XK_KP_Insert, ControlMask, "\033[L", -1, 0},
    {XK_KP_Insert, ControlMask, "\033[2;5~", +1, 0},
    {XK_KP_Insert, XK_ANY_MOD, "\033[4h", -1, 0},
    {XK_KP_Insert, XK_ANY_MOD, "\033[2~", +1, 0},
    {XK_KP_Delete, ControlMask, "\033[M", -1, 0},
    {XK_KP_Delete, ControlMask, "\033[3;5~", +1, 0},
    {XK_KP_Delete, ShiftMask, "\033[2K", -1, 0},
    {XK_KP_Delete, ShiftMask, "\033[3;2~", +1, 0},
    {XK_KP_Delete, XK_ANY_MOD, "\033[P", -1, 0},
    {XK_KP_Delete, XK_ANY_MOD, "\033[3~", +1, 0},
    {XK_KP_Multiply, XK_ANY_MOD, "\033Oj", +2, 0},
    {XK_KP_Add, XK_ANY_MOD, "\033Ok", +2, 0},
    {XK_KP_Enter, XK_ANY_MOD, "\033OM", +2, 0},
    {XK_KP_Enter, XK_ANY_MOD, "\r", -1, 0},
    {XK_KP_Subtract, XK_ANY_MOD, "\033Om", +2, 0},
    {XK_KP_Decimal, XK_ANY_MOD, "\033On", +2, 0},
    {XK_KP_Divide, XK_ANY_MOD, "\033Oo", +2, 0},
    {XK_KP_0, XK_ANY_MOD, "\033Op", +2, 0},
    {XK_KP_1, XK_ANY_MOD, "\033Oq", +2, 0},
    {XK_KP_2, XK_ANY_MOD, "\033Or", +2, 0},
    {XK_KP_3, XK_ANY_MOD, "\033Os", +2, 0},
    {XK_KP_4, XK_ANY_MOD, "\033Ot", +2, 0},
    {XK_KP_5, XK_ANY_MOD, "\033Ou", +2, 0},
    {XK_KP_6, XK_ANY_MOD, "\033Ov", +2, 0},
    {XK_KP_7, XK_ANY_MOD, "\033Ow", +2, 0},
    {XK_KP_8, XK_ANY_MOD, "\033Ox", +2, 0},
    {XK_KP_9, XK_ANY_MOD, "\033Oy", +2, 0},
    {XK_Up, ShiftMask, "\033[1;2A", 0, 0},
    {XK_Up, Mod1Mask, "\033[1;3A", 0, 0},
    {XK_Up, ShiftMask | Mod1Mask, "\033[1;4A", 0, 0},
    {XK_Up, ControlMask, "\033[1;5A", 0, 0},
    {XK_Up, ShiftMask | ControlMask, "\033[1;6A", 0, 0},
    {XK_Up, ControlMask | Mod1Mask, "\033[1;7A", 0, 0},
    {XK_Up, ShiftMask | ControlMask | Mod1Mask, "\033[1;8A", 0, 0},
    {XK_Up, XK_ANY_MOD, "\033[A", 0, -1},
    {XK_Up, XK_ANY_MOD, "\033OA", 0, +1},
    {XK_Down, ShiftMask, "\033[1;2B", 0, 0},
    {XK_Down, Mod1Mask, "\033[1;3B", 0, 0},
    {XK_Down, ShiftMask | Mod1Mask, "\033[1;4B", 0, 0},
    {XK_Down, ControlMask, "\033[1;5B", 0, 0},
    {XK_Down, ShiftMask | ControlMask, "\033[1;6B", 0, 0},
    {XK_Down, ControlMask | Mod1Mask, "\033[1;7B", 0, 0},
    {XK_Down, ShiftMask | ControlMask | Mod1Mask, "\033[1;8B", 0, 0},
    {XK_Down, XK_ANY_MOD, "\033[B", 0, -1},
    {XK_Down, XK_ANY_MOD, "\033OB", 0, +1},
    {XK_Left, ShiftMask, "\033[1;2D", 0, 0},
    {XK_Left, Mod1Mask, "\033[1;3D", 0, 0},
    {XK_Left, ShiftMask | Mod1Mask, "\033[1;4D", 0, 0},
    {XK_Left, ControlMask, "\033[1;5D", 0, 0},
    {XK_Left, ShiftMask | ControlMask, "\033[1;6D", 0, 0},
    {XK_Left, ControlMask | Mod1Mask, "\033[1;7D", 0, 0},
    {XK_Left, ShiftMask | ControlMask | Mod1Mask, "\033[1;8D", 0, 0},
    {XK_Left, XK_ANY_MOD, "\033[D", 0, -1},
    {XK_Left, XK_ANY_MOD, "\033OD", 0, +1},
    {XK_Right, ShiftMask, "\033[1;2C", 0, 0},
    {XK_Right, Mod1Mask, "\033[1;3C", 0, 0},
    {XK_Right, ShiftMask | Mod1Mask, "\033[1;4C", 0, 0},
    {XK_Right, ControlMask, "\033[1;5C", 0, 0},
    {XK_Right, ShiftMask | ControlMask, "\033[1;6C", 0, 0},
    {XK_Right, ControlMask | Mod1Mask, "\033[1;7C", 0, 0},
    {XK_Right, ShiftMask | ControlMask | Mod1Mask, "\033[1;8C", 0, 0},
    {XK_Right, XK_ANY_MOD, "\033[C", 0, -1},
    {XK_Right, XK_ANY_MOD, "\033OC", 0, +1},
    {XK_ISO_Left_Tab, ShiftMask, "\033[Z", 0, 0},
    {XK_Return, Mod1Mask, "\033\r", 0, 0},
    {XK_Return, XK_ANY_MOD, "\r", 0, 0},
    {XK_Insert, ShiftMask, "\033[4l", -1, 0},
    {XK_Insert, ShiftMask, "\033[2;2~", +1, 0},
    {XK_Insert, ControlMask, "\033[L", -1, 0},
    {XK_Insert, ControlMask, "\033[2;5~", +1, 0},
    {XK_Insert, XK_ANY_MOD, "\033[4h", -1, 0},
    {XK_Insert, XK_ANY_MOD, "\033[2~", +1, 0},
    {XK_Delete, ControlMask, "\033[M", -1, 0},
    {XK_Delete, ControlMask, "\033[3;5~", +1, 0},
    {XK_Delete, ShiftMask, "\033[2K", -1, 0},
    {XK_Delete, ShiftMask, "\033[3;2~", +1, 0},
    {XK_Delete, XK_ANY_MOD, "\033[P", -1, 0},
    {XK_Delete, XK_ANY_MOD, "\033[3~", +1, 0},
    {XK_BackSpace, XK_NO_MOD, "\177", 0, 0},
    {XK_BackSpace, Mod1Mask, "\033\177", 0, 0},
    {XK_Home, ShiftMask, "\033[2J", 0, -1},
    {XK_Home, ShiftMask, "\033[1;2H", 0, +1},
    {XK_Home, XK_ANY_MOD, "\033[H", 0, -1},
    {XK_Home, XK_ANY_MOD, "\033[1~", 0, +1},
    {XK_End, ControlMask, "\033[J", -1, 0},
    {XK_End, ControlMask, "\033[1;5F", +1, 0},
    {XK_End, ShiftMask, "\033[K", -1, 0},
    {XK_End, ShiftMask, "\033[1;2F", +1, 0},
    {XK_End, XK_ANY_MOD, "\033[4~", 0, 0},
    {XK_Prior, ControlMask, "\033[5;5~", 0, 0},
    {XK_Prior, ShiftMask, "\033[5;2~", 0, 0},
    {XK_Prior, XK_ANY_MOD, "\033[5~", 0, 0},
    {XK_Next, ControlMask, "\033[6;5~", 0, 0},
    {XK_Next, ShiftMask, "\033[6;2~", 0, 0},
    {XK_Next, XK_ANY_MOD, "\033[6~", 0, 0},
    {XK_F1, XK_NO_MOD, "\033OP", 0, 0},
    {XK_F1, /* F13 */ ShiftMask, "\033[1;2P", 0, 0},
    {XK_F1, /* F25 */ ControlMask, "\033[1;5P", 0, 0},
    {XK_F1, /* F37 */ Mod4Mask, "\033[1;6P", 0, 0},
    {XK_F1, /* F49 */ Mod1Mask, "\033[1;3P", 0, 0},
    {XK_F1, /* F61 */ Mod3Mask, "\033[1;4P", 0, 0},
    {XK_F2, XK_NO_MOD, "\033OQ", 0, 0},
    {XK_F2, /* F14 */ ShiftMask, "\033[1;2Q", 0, 0},
    {XK_F2, /* F26 */ ControlMask, "\033[1;5Q", 0, 0},
    {XK_F2, /* F38 */ Mod4Mask, "\033[1;6Q", 0, 0},
    {XK_F2, /* F50 */ Mod1Mask, "\033[1;3Q", 0, 0},
    {XK_F2, /* F62 */ Mod3Mask, "\033[1;4Q", 0, 0},
    {XK_F3, XK_NO_MOD, "\033OR", 0, 0},
    {XK_F3, /* F15 */ ShiftMask, "\033[1;2R", 0, 0},
    {XK_F3, /* F27 */ ControlMask, "\033[1;5R", 0, 0},
    {XK_F3, /* F39 */ Mod4Mask, "\033[1;6R", 0, 0},
    {XK_F3, /* F51 */ Mod1Mask, "\033[1;3R", 0, 0},
    {XK_F3, /* F63 */ Mod3Mask, "\033[1;4R", 0, 0},
    {XK_F4, XK_NO_MOD, "\033OS", 0, 0},
    {XK_F4, /* F16 */ ShiftMask, "\033[1;2S", 0, 0},
    {XK_F4, /* F28 */ ControlMask, "\033[1;5S", 0, 0},
    {XK_F4, /* F40 */ Mod4Mask, "\033[1;6S", 0, 0},
    {XK_F4, /* F52 */ Mod1Mask, "\033[1;3S", 0, 0},
    {XK_F5, XK_NO_MOD, "\033[15~", 0, 0},
    {XK_F5, /* F17 */ ShiftMask, "\033[15;2~", 0, 0},
    {XK_F5, /* F29 */ ControlMask, "\033[15;5~", 0, 0},
    {XK_F5, /* F41 */ Mod4Mask, "\033[15;6~", 0, 0},
    {XK_F5, /* F53 */ Mod1Mask, "\033[15;3~", 0, 0},
    {XK_F6, XK_NO_MOD, "\033[17~", 0, 0},
    {XK_F6, /* F18 */ ShiftMask, "\033[17;2~", 0, 0},
    {XK_F6, /* F30 */ ControlMask, "\033[17;5~", 0, 0},
    {XK_F6, /* F42 */ Mod4Mask, "\033[17;6~", 0, 0},
    {XK_F6, /* F54 */ Mod1Mask, "\033[17;3~", 0, 0},
    {XK_F7, XK_NO_MOD, "\033[18~", 0, 0},
    {XK_F7, /* F19 */ ShiftMask, "\033[18;2~", 0, 0},
    {XK_F7, /* F31 */ ControlMask, "\033[18;5~", 0, 0},
    {XK_F7, /* F43 */ Mod4Mask, "\033[18;6~", 0, 0},
    {XK_F7, /* F55 */ Mod1Mask, "\033[18;3~", 0, 0},
    {XK_F8, XK_NO_MOD, "\033[19~", 0, 0},
    {XK_F8, /* F20 */ ShiftMask, "\033[19;2~", 0, 0},
    {XK_F8, /* F32 */ ControlMask, "\033[19;5~", 0, 0},
    {XK_F8, /* F44 */ Mod4Mask, "\033[19;6~", 0, 0},
    {XK_F8, /* F56 */ Mod1Mask, "\033[19;3~", 0, 0},
    {XK_F9, XK_NO_MOD, "\033[20~", 0, 0},
    {XK_F9, /* F21 */ ShiftMask, "\033[20;2~", 0, 0},
    {XK_F9, /* F33 */ ControlMask, "\033[20;5~", 0, 0},
    {XK_F9, /* F45 */ Mod4Mask, "\033[20;6~", 0, 0},
    {XK_F9, /* F57 */ Mod1Mask, "\033[20;3~", 0, 0},
    {XK_F10, XK_NO_MOD, "\033[21~", 0, 0},
    {XK_F10, /* F22 */ ShiftMask, "\033[21;2~", 0, 0},
    {XK_F10, /* F34 */ ControlMask, "\033[21;5~", 0, 0},
    {XK_F10, /* F46 */ Mod4Mask, "\033[21;6~", 0, 0},
    {XK_F10, /* F58 */ Mod1Mask, "\033[21;3~", 0, 0},
    {XK_F11, XK_NO_MOD, "\033[23~", 0, 0},
    {XK_F11, /* F23 */ ShiftMask, "\033[23;2~", 0, 0},
    {XK_F11, /* F35 */ ControlMask, "\033[23;5~", 0, 0},
    {XK_F11, /* F47 */ Mod4Mask, "\033[23;6~", 0, 0},
    {XK_F11, /* F59 */ Mod1Mask, "\033[23;3~", 0, 0},
    {XK_F12, XK_NO_MOD, "\033[24~", 0, 0},
    {XK_F12, /* F24 */ ShiftMask, "\033[24;2~", 0, 0},
    {XK_F12, /* F36 */ ControlMask, "\033[24;5~", 0, 0},
    {XK_F12, /* F48 */ Mod4Mask, "\033[24;6~", 0, 0},
    {XK_F12, /* F60 */ Mod1Mask, "\033[24;3~", 0, 0},
    {XK_F13, XK_NO_MOD, "\033[1;2P", 0, 0},
    {XK_F14, XK_NO_MOD, "\033[1;2Q", 0, 0},
    {XK_F15, XK_NO_MOD, "\033[1;2R", 0, 0},
    {XK_F16, XK_NO_MOD, "\033[1;2S", 0, 0},
    {XK_F17, XK_NO_MOD, "\033[15;2~", 0, 0},
    {XK_F18, XK_NO_MOD, "\033[17;2~", 0, 0},
    {XK_F19, XK_NO_MOD, "\033[18;2~", 0, 0},
    {XK_F20, XK_NO_MOD, "\033[19;2~", 0, 0},
    {XK_F21, XK_NO_MOD, "\033[20;2~", 0, 0},
    {XK_F22, XK_NO_MOD, "\033[21;2~", 0, 0},
    {XK_F23, XK_NO_MOD, "\033[23;2~", 0, 0},
    {XK_F24, XK_NO_MOD, "\033[24;2~", 0, 0},
    {XK_F25, XK_NO_MOD, "\033[1;5P", 0, 0},
    {XK_F26, XK_NO_MOD, "\033[1;5Q", 0, 0},
    {XK_F27, XK_NO_MOD, "\033[1;5R", 0, 0},
    {XK_F28, XK_NO_MOD, "\033[1;5S", 0, 0},
    {XK_F29, XK_NO_MOD, "\033[15;5~", 0, 0},
    {XK_F30, XK_NO_MOD, "\033[17;5~", 0, 0},
    {XK_F31, XK_NO_MOD, "\033[18;5~", 0, 0},
    {XK_F32, XK_NO_MOD, "\033[19;5~", 0, 0},
    {XK_F33, XK_NO_MOD, "\033[20;5~", 0, 0},
    {XK_F34, XK_NO_MOD, "\033[21;5~", 0, 0},
    {XK_F35, XK_NO_MOD, "\033[23;5~", 0, 0},
};

/*
 * Selection types' masks.
 * Use the same masks as usual.
 * Button1Mask is always unset, to make masks match between ButtonPress.
 * ButtonRelease and MotionNotify.
 * If no match is found, regular selection is used.
 */
static uint selmasks[] = {
    [SEL_RECTANGULAR] = Mod1Mask,
};

/*
 * Printable characters in ASCII, used to estimate the advance width
 * of single wide characters.
 */
static char ascii_printable[] = " !\"#$%&'()*+,-./0123456789:;<=>?"
                                "@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_"
                                "`abcdefghijklmnopqrstuvwxyz{|}~";
